pym/bob/cmds/build.py

```diff
@@ -980,20 +1008,62 @@ def _cookPackageStep(self, packageStep, checkoutOnly, depth):
+                    # copy to shared location if possible
+                    self.__installSharedPackage(packageBuildId, packageStep, prettyPackagePath)
                     workspaceChanged = True
                     self.__statistic.packagesBuilt += 1
                     if packageBuildId and self.__archive.canUploadLocal():
-                        self.__archive.uploadPackage(packageBuildId, audit, prettyPackagePath, self.__verbose)
+                        if packageStep.isRelocatable():
+                            self.__archive.uploadPackage(packageBuildId, audit, prettyPackagePath, self.__verbose)
+
 
             # Rehash directory if content was changed
             if workspaceChanged:
                 BobState().setResultHash(prettyPackagePath, packageHash)
-                if wasDownloaded:
+                if wasDownloadedOrShared:
                     BobState().setInputHashes(prettyPackagePath, packageBuildId)
                 else:
                     BobState().setInputHashes(prettyPackagePath, [packageBuildId] + packageInputHashes)
             self._setAlreadyRun(packageStep, False, checkoutOnly)
 
+    def __linkSharedPackage(self, packageBuildId, step, workspace):
+        if step.isShared() and step.isRelocatable() and self.__sharedFolder and self._checkSharedPackage(asHexStr(packageBuildId)):
+            sharedpackagepath = os.path.join(self.__sharedFolder, asHexStr(packageBuildId))
+            print(colorize("     Link from shared location {}".format(sharedpackagepath), "32"))
+            dirlist = os.listdir(sharedpackagepath)
+            for item in dirlist:
+                itempath = os.path.join(sharedpackagepath, item)
+                if item in ['audit.json.gz']:
+                    os.symlink(itempath, os.path.join(workspace, '..', item))
+                else:
+                    os.symlink(itempath, os.path.join(workspace, item), os.path.isdir(itempath))
+            return True
+        return False
+
+    def __installSharedPackage(self, packageBuildId, step, workspace):
+        if step.isShared() and step.isRelocatable() and self.__sharedFolder and not self._checkSharedPackage(asHexStr(packageBuildId)):
+            buildidhexstr = asHexStr(packageBuildId)
+            sharedpackagepath = os.path.join(self.__sharedFolder, buildidhexstr)
+            print(colorize("     Install in shared location {}".format(sharedpackagepath), "32"))
+            # create temp dir
+            try:
+                tmpdir = tempfile.mkdtemp(dir=self.__sharedFolder)
+            except OSError:
+                raise OSError("Temporary directory couldn't be created.")
+            tmppackagepath = os.path.join(tmpdir, buildidhexstr)
+            # copy all files from workspace - keep symlinks
+            shutil.copytree(workspace, tmppackagepath, symlinks=True)
+            # copy audit
+            shutil.copy2(os.path.join(workspace, '..', 'audit.json.gz'), tmppackagepath)
```

Comment:
The audit trail is stored together with the results? What would happen if the package result contains a "audit.json.gz"? Better store the "workspace" and "audit.json.gz" separately.